src/nemos/solvers.py

```diff
@@ -0,0 +1,725 @@
+from functools import partial
+from typing import Any, Callable, NamedTuple, Optional, Union
+
+import jax
+import jax.flatten_util
+import jax.numpy as jnp
+from jax import grad, jit, lax, random
+from jaxopt import OptStep
+from jaxopt._src import loop
+from jaxopt.prox import prox_none
+
+from .tree_utils import tree_add_scalar_mul, tree_l2_norm, tree_slice, tree_sub
+from .typing import KeyArrayLike, Pytree
+
+
+class SVRGState(NamedTuple):
```

Comment:
should we be using a NamedTuple for this or a [dataclass](https://docs.python.org/3/library/dataclasses.html)?

From [this stackoverflow](https://stackoverflow.com/questions/51671699/data-classes-vs-typing-namedtuple-primary-use-cases), it looks like NamedTuples make sense when the data structure is immutable, but the state does get changed, right?

Reply:
The state does get updated, but by creating a new one (`._replace` creates a copy) instead of modifying the original one.
I basically just copied [how jaxopt defines optimizer states](https://github.com/google/jaxopt/blob/6d882923a54feb6f68b46193d9787cc73bfdadd3/jaxopt/_src/proximal_gradient.py#L84), which might have been inspired by [this example](https://jax.readthedocs.io/en/latest/stateful-computations.html#simple-worked-example-linear-regression).